skimage/registration/_affine.py

```diff
@@ -0,0 +1,201 @@
+import functools
+import numpy as np
+from scipy import ndimage as ndi
+from scipy.optimize import minimize
+
+from skimage.transform.pyramids import pyramid_gaussian
+from skimage.metrics import normalized_mutual_information
+
+__all__ = ['affine']
+
+
+def _parameter_vector_to_matrix(parameter_vector):
+    """Convert m optimization parameters to a (n+1, n+1) transformation matrix.
+
+    By default (the case of this function), the parameter vector is taken to
+    be the first n rows of the affine transformation matrix in homogeneous
+    coordinate space.
+
+    Parameters
+    ----------
+    parameter_vector : (ndim*(ndim+1)) array
+        A vector of M = N * (N+1) parameters.
+
+    Returns
+    -------
+    matrix : (ndim+1, ndim+1) array
+        A transformation matrix used to affine-map coordinates in an
+        ``ndim``-dimensional space.
+    """
+    m = parameter_vector.shape[0]
+    ndim = int((np.sqrt(4*m + 1) - 1) / 2)
+    top_matrix = np.reshape(parameter_vector, (ndim, ndim+1))
+    bottom_row = np.array([[0] * ndim + [1]])
+    return np.concatenate((top_matrix, bottom_row), axis=0)
+
+
+def cost_nmi(image0, image1, *, bins=100):
+    """Negative of the normalized mutual information.
+
+    See :func:`skimage.metrics.normalized_mutual_information` for more info.
+
+    Parameters
+    ----------
+    image0, image1 : array
+        The images to be compared. They should have the same shape.
+    bins : int or sequence of int, optional
+        The granularity of the histogram with which to compare the images.
+        If it's a sequence, each number is the number of bins for that image.
+
+    Returns
+    -------
+    cnmi : float
+        The negative of the normalized mutual information between ``image0``
+        and ``image1``.
+    """
+    return -normalized_mutual_information(image0, image1, bins=bins)
+
+
+def _param_cost(reference_image, moving_image, parameter_vector, *,
+                vector_to_matrix, cost, multichannel):
+    transformation = vector_to_matrix(parameter_vector)
+    if not multichannel:
+        transformed = ndi.affine_transform(moving_image, transformation,
+                                           order=1)
+    else:
+        transformed = np.zeros_like(moving_image)
+        for ch in range(moving_image.shape[-1]):
+            ndi.affine_transform(moving_image[..., ch], transformation,
+                                 order=1, output=transformed[..., ch])
+    return cost(reference_image, transformed)
+
+
+def affine(reference_image, moving_image, *,
+           cost=cost_nmi, initial_vector=None,
+           translation_indices=None, vector_to_matrix=None,
+           pyramid_scale=2, pyramid_minimum_size=8, multichannel=False,
+           inverse=True, level_callback=None, method='Powell',
+           **kwargs):
+    """Find a transformation matrix to register a moving image to a reference.
+
+    Parameters
```

Comment:
The parameters seem to be ordered in reverse order of importance, going from technical things the user typically don't care about, to very important parameters that control the registration; shall we reshuffle?

Reply:
I'm certainly up for reshuffling! But I think inverse order isn't quite right. 😂 I would say the cost function is definitely the most critical parameter, and level_callback is among the least. The optimisation method... Yeah, that can probably be moved up. =)

Anyway, I'm very happy to contemplate alternate orderings!

Reply:
Suggested ordering: cost function, optimization parameters, pyramid parameters, rest.

```
reference_image, moving_image,
*,
cost='nmi',
method='Powell',
initial_parameters=None,
pyramid_scale=2,
pyramid_minimum_size=8,
level_callback=None,
inverse=True,
multichannel=False
```

Not sure where `multichannel` should go.  Default value for `pyramid_minimum_size` feels too small.

Did we experiment with different default optimizers?

Reply:
> Did we experiment with different default optimizers

Only briefly. Maybe worth revisiting.

> Default value for pyramid_minimum_size feels too small.

Unfortunately the registration is quite sensitive to this parameter. The first registration in the example fails without this. The second one fails without 32 being the limit.

Reply:
I'll play around with different optimisers using a more reasonable 16 or 32 minimum size.

Reply:
@stefanv given that Powell is the *only* method currently that does anything useful (see comment below), I'd argue that maybe the optimization method shouldn't be up top like that...

Also note, I have *not* looked at basin-hopping.

Reply:
Sure, that one can move down.

Also, we have one, maybe two, test cases—so we shouldn't make assumptions about robustness based on such limited testing.

Reply:
Basin hopping seemed to solve so many problems for us in Elegant SciPy; I hope we do add it!

Reply:
The other parameter I was missing: interpolation order.  Can be very impactful, both in terms of speed and accuracy.